db/migrate/20200813051337_add_private_message_to_post_search_data.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class AddPrivateMessageToPostSearchData < ActiveRecord::Migration[6.0]
+  def up
+    add_column :post_search_data, :private_message, :boolean
+
+    # Delete post_search_data of orphaned posts
+    execute <<~SQL
+    DELETE FROM post_search_data
+    WHERE post_id IN (
+      SELECT posts.id
+      FROM posts
+      LEFT JOIN topics ON topics.id = posts.topic_id
+      WHERE topics.id IS NULL
+    )
+    SQL
+
+    # Delete orphaned post_search_data
+    execute <<~SQL
+    DELETE FROM post_search_data
+    WHERE post_id IN (
+      SELECT post_search_data.post_id
+      FROM post_search_data
+      LEFT JOIN posts ON posts.id = post_search_data.post_id
+      WHERE posts.id IS NULL
+    )
+    SQL
+
+    execute <<~SQL
+    UPDATE post_search_data
+    SET private_message = true
+    FROM posts
+    INNER JOIN topics ON topics.id = posts.topic_id AND topics.archetype = 'private_message'
+    WHERE posts.id = post_search_data.post_id
+    SQL
+
+    execute <<~SQL
```

Comment:
If you add a default value of `false` to the new column you could create the `not null` index above, avoid the second query below, and the changing of column to not null.

Reply:
This is intentional because I want the migration the fail if any `post_seach_data` rows do not have the `private_message` column set to `false` even after the queries to migrate the data has been written. There is a risk that posts may incorrectly show up in search if we default the column to `false`.